[Cocoa] Adopt CASpatialAudioExperience for WebAudio

jernoble opened the pull request:
#### e69c00275149ab466509f9d5cb8a8634bc80e7a5
<pre>
[Cocoa] Adopt CASpatialAudioExperience for WebAudio
<a href="https://bugs.webkit.org/show_bug.cgi?id=288249">https://bugs.webkit.org/show_bug.cgi?id=288249</a>
<a href="https://rdar.apple.com/145326961">rdar://145326961</a>

Reviewed by Andy Estes.

Add support for creating a CASpatialAudioExperience and attaching that
experience to the CoreAudio AudioOutputUnit used by WebAudio.

The number of parameters taken by various AudioDestination subclasses
and strategies has become burdensome, so rather than add an additional
parameter needed for creating a CASpatialAudioExperience, add a new
AudioDestinationCreationOptions struct and use that struct for the
same purposes.

This new AudioDestinationCreationOptions struct will also contain a
sceneIdentifier used to anchor audio generation to a particular screen
location.

However, most audio on Cocoa ports flows through SharedAudioDestination,
which muxes that audio together into a single AudioUnit. Add the
spatialIdentifier as a key in the SharedAudioDestination&apos;s internal hash
map, and re-create the destinations adaptor when that destination&apos;s
sceneIdentifier changes.

The sceneIdentifier and preference must be passed through to the GPU
process so that it can be accessed by RemoteAudioDestinationManager.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::sceneIdentifierDidChange):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::createDestination):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::inputDeviceId const):
(WebCore::AudioDestination::numberOfInputChannels const):
(WebCore::AudioDestination::numberOfOutputChannels const):
(WebCore::AudioDestination::sceneIdentifier const):
(WebCore::AudioDestination::setSceneIdentifier):
(WebCore::AudioDestination::AudioDestination):
* Source/WebCore/platform/audio/AudioDestinationResampler.cpp:
(WebCore::AudioDestinationResampler::AudioDestinationResampler):
* Source/WebCore/platform/audio/AudioDestinationResampler.h:
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::sceneIdentifier const):
(WebCore::SharedAudioDestinationAdapter::takeEnsureFunction):
(WebCore::SharedAudioDestinationAdapter::ensureAdapter):
(WebCore::SharedAudioDestinationAdapter::SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestination::create):
(WebCore::SharedAudioDestination::SharedAudioDestination):
(WebCore::SharedAudioDestination::setSceneIdentifier):
* Source/WebCore/platform/audio/SharedAudioDestination.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestination::create):
(WebCore::AudioDestinationCocoa::AudioDestinationCocoa):
(WebCore::AudioDestinationCocoa::setSceneIdentifier):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.cpp:
(WebCore::AudioOutputUnitAdaptor::setSpatialAudioExperience):
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.h:
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.mm:
(WebCore::createSpatialAudioExperienceWithOptions):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::sceneIdentifierDidChange):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/mediastream/AudioMediaStreamTrackRenderer.h:
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.cpp:
(WebCore::AudioTrackPrivateMediaStream::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererCocoa.cpp:
(WebCore::AudioMediaStreamTrackRendererCocoa::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererCocoa.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.h:
(WebCore::AudioMediaStreamTrackRendererInternalUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
(WebCore::AudioMediaStreamTrackRendererUnit::Unit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.h:
* Source/WebCore/platform/mediastream/cocoa/BaseAudioMediaStreamTrackRendererUnit.h:
(WebCore::BaseAudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/IncomingAudioMediaStreamTrackRendererUnit.cpp:
(WebCore::IncomingAudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/IncomingAudioMediaStreamTrackRendererUnit.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.cpp:
(WebCore::MockAudioDestinationCocoa::MockAudioDestinationCocoa):
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
(WebKit::RemoteAudioDestinationManager::setSceneIdentifier):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::create):
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::setSceneIdentifier):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:

Canonical link: <a href="https://commits.webkit.org/291404@main">https://commits.webkit.org/291404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/435bcbd0eef3e661c403931811757cd4f74bf7e9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92695 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97686 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43206 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94745 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12521 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20698 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70979 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28416 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95697 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9484 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51310 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9176 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42534 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85405 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79489 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99710 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91361 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19747 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14538 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79987 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79289 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19671 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-conditional/container-queries/canvas-as-container-001.html (failure)") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23811 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1135 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12761 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14827 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19731 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24907 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114009 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19418 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32940 "Found 10 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.default, wasm.yaml/wasm/spec-tests/float_exprs.wast.js.wasm-collect-continuously, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-collect-continuously, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch-with-delegate.js.wasm-slow-memory, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22878 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
```diff
@@ -116,12 +117,11 @@ void DefaultAudioDestinationNode::createDestination()
-        m_destination->setSceneIdentifier(String { context().sceneIdentifier() });
+        , context().sceneIdentifier()
 #endif
+        });
```
```suggestion
    });
```

aestes commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -57,12 +58,22 @@ class SharedAudioDestinationAdapter : public ThreadSafeRefCountedAndCanMakeThrea
+    const String& sceneIdentifier() const { return m_sceneIdentifier; }
+#endif
+
+    AudioDestinationCreationFunction takeEnsureFunction() { return WTFMove(m_ensureFunction); }
```
```suggestion
    AudioDestinationCreationFunction takeEnsureFunction() { return std::exchange(m_ensureFunction, nullptr); }
```

aestes commented on Source/WebCore/platform/audio/SharedAudioDestination.cpp:
```diff
@@ -98,32 +114,45 @@ auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+#if PLATFORM(IOS_FAMILY)
+        , options.sceneIdentifier.isNull() ? emptyString() : options.sceneIdentifier
+#endif
+        }) }
```
```suggestion
    }) }
```

jernoble committed:
[Cocoa] Adopt CASpatialAudioExperience for WebAudio
https://bugs.webkit.org/show_bug.cgi?id=288249
rdar://145326961

Reviewed by Andy Estes.

Add support for creating a CASpatialAudioExperience and attaching that
experience to the CoreAudio AudioOutputUnit used by WebAudio.

The number of parameters taken by various AudioDestination subclasses
and strategies has become burdensome, so rather than add an additional
parameter needed for creating a CASpatialAudioExperience, add a new
AudioDestinationCreationOptions struct and use that struct for the
same purposes.

This new AudioDestinationCreationOptions struct will also contain a
sceneIdentifier used to anchor audio generation to a particular screen
location.

However, most audio on Cocoa ports flows through SharedAudioDestination,
which muxes that audio together into a single AudioUnit. Add the
spatialIdentifier as a key in the SharedAudioDestination's internal hash
map, and re-create the destinations adaptor when that destination's
sceneIdentifier changes.

The sceneIdentifier and preference must be passed through to the GPU
process so that it can be accessed by RemoteAudioDestinationManager.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::sceneIdentifierDidChange):
* Source/WebCore/Modules/webaudio/DefaultAudioDestinationNode.cpp:
(WebCore::DefaultAudioDestinationNode::createDestination):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::sceneIdentifierDidChange):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/audio/AudioDestination.h:
(WebCore::AudioDestination::inputDeviceId const):
(WebCore::AudioDestination::numberOfInputChannels const):
(WebCore::AudioDestination::numberOfOutputChannels const):
(WebCore::AudioDestination::sceneIdentifier const):
(WebCore::AudioDestination::setSceneIdentifier):
(WebCore::AudioDestination::AudioDestination):
* Source/WebCore/platform/audio/AudioDestinationResampler.cpp:
(WebCore::AudioDestinationResampler::AudioDestinationResampler):
* Source/WebCore/platform/audio/AudioDestinationResampler.h:
* Source/WebCore/platform/audio/SharedAudioDestination.cpp:
(WebCore::SharedAudioDestinationAdapter::sceneIdentifier const):
(WebCore::SharedAudioDestinationAdapter::takeEnsureFunction):
(WebCore::SharedAudioDestinationAdapter::ensureAdapter):
(WebCore::SharedAudioDestinationAdapter::SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestinationAdapter::~SharedAudioDestinationAdapter):
(WebCore::SharedAudioDestination::create):
(WebCore::SharedAudioDestination::SharedAudioDestination):
(WebCore::SharedAudioDestination::setSceneIdentifier):
* Source/WebCore/platform/audio/SharedAudioDestination.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestination::create):
(WebCore::AudioDestinationCocoa::AudioDestinationCocoa):
(WebCore::AudioDestinationCocoa::setSceneIdentifier):
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.h:
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.cpp:
(WebCore::AudioOutputUnitAdaptor::setSpatialAudioExperience):
* Source/WebCore/platform/audio/cocoa/AudioOutputUnitAdaptor.h:
* Source/WebCore/platform/audio/cocoa/SpatialAudioExperienceHelper.mm:
(WebCore::createSpatialAudioExperienceWithOptions):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::sceneIdentifierDidChange):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::updateSpatialTrackingLabel):
* Source/WebCore/platform/mediastream/AudioMediaStreamTrackRenderer.h:
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.cpp:
(WebCore::AudioTrackPrivateMediaStream::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/AudioTrackPrivateMediaStream.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererCocoa.cpp:
(WebCore::AudioMediaStreamTrackRendererCocoa::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererCocoa.h:
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.cpp:
(WebCore::LocalAudioMediaStreamTrackRendererInternalUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererInternalUnit.h:
(WebCore::AudioMediaStreamTrackRendererInternalUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.cpp:
(WebCore::AudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
(WebCore::AudioMediaStreamTrackRendererUnit::Unit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/AudioMediaStreamTrackRendererUnit.h:
* Source/WebCore/platform/mediastream/cocoa/BaseAudioMediaStreamTrackRendererUnit.h:
(WebCore::BaseAudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/IncomingAudioMediaStreamTrackRendererUnit.cpp:
(WebCore::IncomingAudioMediaStreamTrackRendererUnit::setSpatialAudioExperience):
* Source/WebCore/platform/mediastream/cocoa/IncomingAudioMediaStreamTrackRendererUnit.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.cpp:
(WebCore::MockAudioDestinationCocoa::MockAudioDestinationCocoa):
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp:
(WebKit::RemoteAudioDestinationManager::createAudioDestination):
(WebKit::RemoteAudioDestinationManager::setSceneIdentifier):
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.h:
* Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.cpp:
(WebKit::RemoteAudioDestinationProxy::create):
(WebKit::RemoteAudioDestinationProxy::RemoteAudioDestinationProxy):
(WebKit::RemoteAudioDestinationProxy::connection):
(WebKit::RemoteAudioDestinationProxy::setSceneIdentifier):
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.mm:

Canonical link: https://commits.webkit.org/291404@main
